### PR TITLE
Adds duration sampler distinct from latency in supplying two bounds

### PR DESCRIPTION
**Pull request by garry-cairns**
**Description:** Adds a bounded duration sampling processor, distinct from the existing latency one in that it has both lower and upper bounds

Apologies for this appearing as a pull request out of nothing, my intent had actually been to create a review area against my own fork and raise an issue asking if you'd accept the PR. I think the need here is pretty obvious from the context, though I think it's easy to imagine preferring this to be a change to the existing processor. I raised as a new one as I thought it might make existing behavior cleaner to retain.

**Link to tracking Issue:** As above this is a bit of a premature PR since I intended to raise as an issue, and thus there isn't one, but I think it's easy enough to deal with here so leaving open for now and have learned GitHub's ways for the future (I rarely use github).

**Testing:** New module so associated tests are added showing all relevant behavior, and passing.

**Documentation:** Updated README and example config

### CODE REVIEW

**Comment by jpkrohling** [comment]
This is a great idea, but I'd rather have it as a parameter to the current latency policy. Are you OK in implementing it as such?

**Comment by garry-cairns** [comment]
Yep I'll give it a go tomorrow or Wednesday

**Comment by garry-cairns** [comment]
Need to do some work admin before signing the CLA so this will be delayed slightly

**Comment by jpkrohling** [comment]
Once the CLA check is passing, I'll review this one.

**Comment by brooklynrob** [comment]
I've signed/executed the CCLA between JPMC and CNCF and added @garry-cairns to the Schedule A, so from a CCLA perspective this PR should be GTG. @jpkrohling

**Comment by brooklynrob** [comment]
> Updated as per review feedback and I've done my bit for the CLA. I think that'll send a request to my org and they'll okay everything. @jpkrohling

Looks like EasyCLA is showing the green COVERED button [above](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/26115?secureweb=Symphony#issuecomment-1694342208), so hopefully GTG from a CLA perspective on this. @jpkrohling @garry-cairns

File: processor/tailsamplingprocessor/README.md [hunk 1]
```diff
@@ -77,7 +77,7 @@ processors:
           {
             name: test-policy-2,
             type: latency,
-            latency: {threshold_ms: 5000}
+            latency: {threshold_ms: 5000, upper_threshold_ms: 10000}
```
Comment (jpkrohling) [inline processor/tailsamplingprocessor/README.md:80]: Should there be more documentation around this? Perhaps at the description for this policy, close to the top of the readme?

**Review by jpkrohling — APPROVED** [review]
LGTM, the doc could be a bit better for this though.

File: processor/tailsamplingprocessor/README.md (hunk 1 above)
Comment (garry-cairns) [reply processor/tailsamplingprocessor/README.md:80]: I've added a commit for this, which also fixes some lint and adds the necessary changelog file.

**Review by jpkrohling — APPROVED** [review]

**Comment by garry-cairns** [comment]
@jpkrohling Thanks for the approval. I see there are still two failing lint checks. `golint` is passing for me locally so, with this being my first change in the repo, I may need a little hand-holding to get that over the line if you have any time to leave some pointers.

**Comment by jpkrohling** [comment]
Running `make lint` when on the tailsamplingprocessor directory yields this:

```
tailsamplingprocessor/internal/sampling/latency.go:17: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/open-telemetry/opentelemetry-collector-contrib) (gci)
	thresholdMs int64
```

You can run `golangci-lint run --fix`, which will cause this change:

```diff
diff --git a/processor/tailsamplingprocessor/internal/sampling/latency.go b/processor/tailsamplingprocessor/internal/sampling/latency.go
index 20e3710b2f..be87f47165 100644
--- a/processor/tailsamplingprocessor/internal/sampling/latency.go
+++ b/processor/tailsamplingprocessor/internal/sampling/latency.go
@@ -14,7 +14,7 @@ import (
 
 type latency struct {
        logger           *zap.Logger
-       thresholdMs int64
+       thresholdMs      int64
        upperThresholdMs int64
 }
 

```

I believe this should be enough to have the build fixed.

**Comment by garry-cairns** [comment]
Perfect, thanks. I'll commit that tonight or tomorrow morning, and send a push after some manual testing.

**Comment by garry-cairns** [comment]
@jpkrohling Do you need anything more from me on this one?

**Comment by jpkrohling** [comment]
I triggered the tests again, if the build is green I'll merge this.

**Comment by brooklynrob** [comment]
Great stuff!

**Pull request merged**